Automattic/jetpack
# Reconnect Notice: In-place Reconnect

#### Changes proposed in this Pull Request:
This PR introduces a reconnection flow, which allows users to reconnect Jetpack in-place, in case of connection errors.
In particular, this PR:
- Adds a `connection/reconnect` endpoint, that accepts `POST` requests, with a **required** `action` parameter. The `action` parameter for now can be only `reconnect`. This corresponds to re-registering the site and deleting all tokens (user and blog), so the users need to re-authorize wpcom and re-issue their blog tokens. In the future, this endpoint will also support the actions: `refresh_blog_token` and `refresh_user_token`. 
- Displays an error notice, in cases of connection errors, along with a `Restore connection` action prompt, that triggers the reconnection flow
- Displays the authorization iframe so that a user can re-connect their wpcom account in-place, as part of the reconnection flow (after successful response from the `connection/reconnect` endpoint the user needs to re-connect because the corresponding user tokens are now deleted). By setting the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME` constant  `true` or using **Safari** browser, instead of displaying the authorization iframe the user is redirected to wpcom in order to authorize their account.
- Moves all related Admin page files to using Core translation functions in the context of [Admin Page: Move from i18n-calypso to Core functions to handle translations](https://github.com/Automattic/jetpack/issues/16481)

#### Jetpack product discussion
`p9dueE-1CF-p2`

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:

**Prerequisites:**
- A "broken" Jetpack connection: Using the Broken Token plugin, invalidate the **blog token**.

**Testing:**
- Make sure after "breaking" the blog token, the connection error is visible along with the `Restore connection` CTA.
- Click the `Restore connection` CTA: The admin UI should blur and you should see an info message that reconnection is in progress
- The authorization iframe should be displayed in order to re-connect your wpcom account
- After re-connecting your wpcom account the page will refresh and your connection should be now healthy (the connection error shouldn't display anymore)

**Note:** By setting the `JETPACK_SHOULD_NOT_USE_CONNECTION_IFRAME` constant  `true` or testing this with **Safari** browser, instead of seeing the authorization iframe you should be redirected to wpcom in order to authorize your account.

**Before:**
N/A

**After:**
https://d.pr/v/oN8gUa

#### Proposed changelog entry for your changes:
* Easy reconnect in case of connection errors.
